lib/tests/triples.nix: add test suite and fix our parser so it passes

amjoseph-nixpkgs opened the pull request:
As promised; sorry it took so long.  Probably best reviewed one commit at a time.

Closes NixOS/nixpkgs#165836

Closes NixOS/nixpkgs#230160

###### Description of changes

This PR adds `lib/tests/triples.nix`, which exhaustively tests that our platform triple parse-then-unparse roundtrip agrees with `gnu-config` for all inputs that it accepts (it may -- and does -- reject triples which `gnu-config` accepts).

This PR also includes the commits needed in order to get the test suite to *pass*.

There are only two major exceptions in nixpkgs to agreement with `gnu-config` -- these are triples which we accept but `gnu-config` rejects:

1. `vc4-elf` (for `vc4-{gcc,binutils}`, reverse engineered fork for broadcom's gpu)
2. ~`javascript-unknown-ghcjs` (for `ghc`)~ fixed upstream 6c78704d542cebfb56d17474fe9f8395e9defb94

There is one minor exception: we canonicalize `""-linux` to `unknown-linux` rather than `pc-linux` as `gnu-config` does in some cases.


###### Things done

- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

CC: @sternenseemann @alyssais @Ericson2314

Commit:
lib/systems: replicate cygwin/msvc parsing hack when unparsing

For almost a decade nixpkgs has parsed `*-*-{cygwin,msvc}` as being
ABIs for a common kernel (windows), rather than as kernels.  It
isn't really reasonable to change nixpkgs' handling at this point,
so this commit replicates the special-case hack when unparsing, in
order to ensure that `(unparse . parse)==id`.

Commit:
lib/systems: accomodate nonstandard `vc4-elf` triple

Since 4aa1ffae041bb9c65eb3067e9dbaaa70710ed100 we have been shipping
reverse-engineered out-of-tree forks of gcc and binutils for the
Broadcom VC4.  Upstream for those forks chose the nonstandard
`vc4-elf` triple to identify their system.  GNU config rejects this
triple (and all `vc4-*` triples).

This commit deals with the fallout from that.

Mainly this commit ensures that we are consistent in using `vc4-elf`
instead of `vc4-none`, and makes sure that parsing->unparsing
round-trips properly.

Commit:
lib/systems: handle Solaris properly

GNU config (correctly) considers Solaris to be a BSD, and therefore
appends its version to the kernel name, which should be `solaris2`
not `solaris`.

This commit updates `lib/systems` to do the same.

Commit:
lib/systems: add TODO regarding execFormats

Commit:
lib/systems: add and check valid kernels for each abi

Each kernel supports only a subset of the universe of possible ABIs.
This commit explicitly lists (opt-in) the allowable abis for each
kernel.  Attempting to parse a kernel with an abi which is not on
the list will fail.  This "search space pruning" is essential to
getting our gnu-config agreement tests down to a manageable set of
test cases.  It also lets us reject nonsense triples instead of
trying to handle them exactly the same way gnu-config does.

Commit:
lib/tests/triples.nix: add gnu-config conformance test suite

This commit adds lib/tests/triples.nix, which exhaustively tests
that our platform triple parse-then-unparse round-trip agrees with
gnu-config for all inputs that it accepts (it may -- and does --
reject triples which `gnu-config` accepts).

Commit:
lib/systems: factor tripleFromSkeleton out of tripleFromSystem

tripleFromSkeleton contains the same logic found in
tripleFromSystem, but is capable of operating on unvalidated
attrsets-of-strings.

Commit:
lib/systems: move powerpc64-be ABI-defaulting out of the parser

In PR #182807 we decided that big-endian PowerPC64 should default to
the new elfv2 ABI.  The implementation in that PR did the defaulting
by adding two lines in the *triple parser*, which is turning out to
be problematic; it means that our parse-then-unparse roundtrip
disagrees with gnu-config.

This commit therefore reverts just those two lines.  If the
defaulting logic needs to be moved elsewhere that should be done.

https://github.com/NixOS/nixpkgs/pull/182807/files#r1234650738

Commit:
lib/systems: default arm-apple-linux to -gnu ABI

This is required for agreement with gnu-config.

Commit:
lib/systems: add vendors.{ibm,xilinx}

Commit:
lib/systems: fix vendor fallback logic

This commit fixes the logic for handling missing/unknown vendor
fields in triples, and extends it to handle mmix and microblaze
CPU-types.

Commit:
lib/systems: omit optExecFormat from triples

gnu-config doesn't include this.

Commit:
lib/systems: rename vendors.none -> vendors.""

gnu-config does not recognize `none` as a vendor -- that string
describes a *kernel*

Commit:
lib/systems: introduce tripleFromSystemLossy

Unfortunately gnu-config triples are, in some rare cases, sensitive
to whether the vendor was specified as `unknown` or was omitted.  In
other words, there are situations where these are not handled
exactly the same way:

- `mips-linux-gnu`
- `mips-unknown-linux-gnu`

This commit introduces `lib.systems.tripleFromSystemLossy`, which
returns `"mips-unknown-linux-gnu"` for both of the platforms above.
This is almost always what you want to use.

Commit:
lib/systems: fix abi inference for aarch32-linux

Commit:
lib/systems: add vendors.solo5

With this commit, we now parse or reject the following identically
to gnu-config:

- aarch64-solo5-none (accept)
- aarch64-solo5-none-elf (reject)

Closes #165836.

Commit:
lib/systems: add kernels."", distinct from kernels.none

This commit adds support for triples with a missing kernel, rather
than kernel="none".  It also adds a parser case for triples which
have a vendor and abi, but no kernel.

This allows to parse "*-unknown-elf" triples.

Closes #230160.

amjoseph-nixpkgs commented on lib/tests/triples.nix:
```diff
@@ -0,0 +1,157 @@
+# rare and unusual cases where nixpkgs has carefully and
+# deliberately decided to diverge from it.
+#
+# We use `gnu-config` as our source of ground truth not because it
+# is elegant (it isn't) or sensible (it isn't), but because it:
+#
+# 1. has an executable (i.e. unambiguous) specification
+# 2. covers a superset of the systems on which nixpkgs could be useful
+# 3. is *not* under nixpkgs' control (bickering/drama avoidance)
+# 4. emanates from an organization (GNU) which understands that
+#    the taxonomy is not merely an implementation detail of one of
+#    their software products, and has demonstrated this
+#    understanding over an extended period of time (multiple
+#    decades).
```
Review thread for discussion of this comment.

RaitoBezarius commented on lib/systems/parse.nix:
```diff
@@ -403,6 +404,15 @@ rec {
   mkSystem = components:
     assert types.parsedPlatform.check components;
+    assert ((components.kernel == kernels.ghcjs) != (components.cpu == cpuTypes.javascript))
+           -> throw "cpu==\"javascript\" and kernel==\"ghcjs\" are valid only with each other; you attempted to use cpu==\"${components.cpu.name}\" with kernel==\"${components.kernel.name}\"";
```
Please use `''` here

amjoseph-nixpkgs replied:
Ah, good call: f1bdbaf9620f82cff0a5bf2260b2c117463a1165

RaitoBezarius commented on lib/systems/parse.nix:
```diff
@@ -403,6 +404,15 @@ rec {
+    assert !(isVc4 components -> (with components; vendor.name == "" && kernel.name == "" && abi.name == "elf"))
+            -> throw (with components; "Broadcom VC4 CPUs may be used only in the grandfathered nonstandard triple `vc4-elf`; you tried to create ${cpu.name}-${vendor.name}-${kernel.name}-${abi.name}");
+    assert !(isJavaScript components -> (with components; (vendor.name == "unknown" || vendor.name == "") && kernel == kernels.ghcjs && abi == abis.unknown))
+            -> throw (with components; "The special `javascript` \"cpu\" may be used only in the grandfathered nonstandard triple `javascript-unknown-ghcjs`; you tried to create ${cpu.name}-${vendor.name}-${kernel.name}-${abi.name}");
```
Please use `''×`

amjoseph-nixpkgs replied:
Ah, good call: f1bdbaf9620f82cff0a5bf2260b2c117463a1165

Ericson2314 commented:
So...do we necessary want to be more like GNU config than other things? Starting with the first commit, LLVM and Rust accept the Windows the 4-part way, and I thought that was better, so I went with that.

We could instead modify GNU Config to accept that too, indeed rewrote the thing so getting changes merged for it is indeed possible.

Ericson2314 commented:
https://github.com/Ericson2314/gnu-config/commit/windows Here is support for `windows-msvc` and `windows-gnu`.

Ericson2314 commented:
Email is https://lists.gnu.org/archive/html/config-patches/2023-06/msg00005.html

Message ID (for non-subscribers to respond) is `<20230626235657.645175-1-list@JohnEricson.me>`.

Ericson2314 commented:
> Unfortunately gnu-config triples are, in some rare cases, sensitive to whether the vendor was specified as `unknown` or was omitted.

This, and the "empty string" ones also look to me like GNU config bugs, it would be good to fix those too.

Commit:
implement formatting suggestion from @RaitoBezarius

amjoseph-nixpkgs commented:
> We could instead modify GNU Config to accept that too

:+1:

> Email is https://lists.gnu.org/archive/html/config-patches/2023-06/msg00005.html

The true test of this PR will be when your `gnu-config` patch gets merged and the version bump of `pkgs.gnu-config` causes a test failure (which tells us to update the test!)

> This, and the "empty string" ones also look to me like GNU config bugs, it would be good to fix those too.

Probably.  To cover the most vexing cases that @sternenseemann found in #165836, would be:

1. If `gnu-config` already accepts `${cpu}-${kernel}-{environment}` then it must also accept `${cpu}-unknown-${kernel}-{environment}`
2. If `gnu-config` already accepts `${cpu}-${vendor}-{environment}` (and parses it as a kernel-less triple) then it must also accept `${cpu}-${vendor}-none-{environment}`

The question is what to do about canonicalization.  In order to truly eliminate the "empty string" we would have to *change* the canonicalization of `${cpu}-${kernel}-{environment}` so that

```
# before change
$ config.sub riscv64-none-elf
riscv64-none-elf

# after change
$ config.sub riscv64-none-elf
riscv64-unknown-none-elf
```

... but I'm uneasy about foisting that kind of change on people who are already using these triples.  It might even be against `gnu-config`'s policies (which would be sensible).

So I don't think we can ever truly get rid of the empty-string coordinates, but we can at least make them less confusing by ensuring that you're never *forced* to omit a coordinate.  But `gnu-config` already treats some missing-coordinate triples as canonical, and changing that might cause more breakage than it's worth.

Ericson2314 commented:
The GNU Config patch is now merged!

> The true test of this PR will be when your `gnu-config` patch gets merged and the version bump of `pkgs.gnu-config` causes a test failure (which tells us to update the test!)

Well the issue is making our stuff more lenient in the meantime. I am find having our test be a "living" thing, but wary of making the Nixpkgs version behave "worse" temporarily or permanently.

Ericson2314 commented:
@amjoseph-nixpkgs For the empty string bit, I would like to take a step back. I think the most realistic goal is *not*

> Nixpkgs and GNU config agree on normalizations

but

> Nixpkgs normal forms are also GNU config normal forms

This allows:

1. GNU config might normalize different for compatibility reasons
2. GNU config might accept things (either at all, or as normal forms) we won't (either at all, or normal forms)

Also long as Nixpkgs only sends GNU config ones that we've normalized (according to our rules), then the above should be sufficient.

Ericson2314 commented:
https://lists.gnu.org/archive/html/config-patches/2023-07/msg00002.html This will make it accept `riscv64-unknown-none-elf`. The trick is `none` can still be parsed as a vendor by default (for backwards compat), but we can *also* accept `none` as a kernel with `elf` as an "OS".

Ericson2314 commented:
https://github.com/Ericson2314/gnu-config/commit/none-elf is where it came from.

Ericson2314 commented:
I don't think it affects this PR, but I did a further cleanup in https://lists.gnu.org/archive/html/config-patches/2023-08/msg00006.html

amjoseph-nixpkgs commented:
> Per [#165836 (comment)](https://github.com/NixOS/nixpkgs/issues/165836#issuecomment-1665956086) the `$cpu-$vendor-none-elf` is now gone in GNU config.

I think you meant to say "`$cpu-$vendor-none-elf` is **now accepted by** GNU config" -- along with `$cpu-none-elf`, which it always accepted.

Meanwhile, this PR has quite some time ago fixed our parser so that *it too* accepts both forms.  So there's no need to change the triples -- the ones we've been using are now accepted by both our parser *and* GNU config.

> Also long as Nixpkgs only sends GNU config ones that we've normalized (according to our rules), then the above should be sufficient.

This is impractical: all of the "Nix doubles" are non-normalized.  As long as Nix uses doubles, we have to normalize them, so we're sensitive to *how* they are normalized.

Ericson2314 commented:
> I think you meant to say "$cpu-$vendor-none-elf is now accepted by GNU config"

Oops, yes I did.

I didn't like the changes to our parser to add more succeeding special cases. I don't like `$cpu-none-elf`. I am fine making it fail to parse rather than parse in a different way, however.

> This is impractical: all of the "Nix doubles" are non-normalized. As long as Nix uses doubles, we have to normalize them, so we're sensitive to how they are normalized.

I meant from the `config` field. For the `system` field GNU config and us already agree, right?

The general principle for me is that it is fine for either us or GNU config to reject something the other accepts, but they shouldn't both accept something and normalize it in a different way.

amjoseph-nixpkgs commented on lib/tests/triples.nix:
```diff
@@ -0,0 +1,157 @@
+      (builtins.tryEval (canonicalize triple))
+        .success))
+
+    #
+    # The following are triples which we have deliberately
+    # chosen to accept differently than gnu-config.  Entries here
+    # impose an ongoing burden on nixpkgs, so PRs which add new
+    # triples here should allow for a longer review period and
+    # should be advertised widely.
+    #
+    (lib.filter (v:
+
+      # gnu-config infers the vendor "pc" for these Nix doubles, but
+      # nixpkgs has used ${cpu}-unknown-linux-gnu for a very long time.
+      v != "x86_64-linux" && v != "i686-linux" &&
+      v != "x86_64-linux-gnu" && v != "i686-linux-gnu" &&
+      v != "x86_64-darwin" && v != "aarch64-darwin" &&
+
+      # We have been shipping Broadcom VC4 forks of gcc/binutils
+      # using nonstandard triples since 4aa1ffae041bb9c65eb3067e9dbaaa70710ed100
+      (canonicalize v) != "vc4-elf"
+
+      # This triple is special to GHC/Cabal/GHCJS and not recognized by autotools
+      # See: https://gitlab.haskell.org/ghc/ghc/-/commit/6636b670233522f01d002c9b97827d00289dbf5c
+      # https://github.com/ghcjs/ghcjs/issues/53
+      && (canonicalize v) != "javascript-unknown-ghcjs"
```
I probably should have titled this PR "add test suite **and fix our broke-ass parser so it passes the tests**".

This was a lot of work.  Very boring and tedious work.

> The general principle for me is that it is fine for either us or GNU config to reject something the other accepts, but they shouldn't both accept something and normalize it in a different way.

That is what this PR does, with five clearly enumerated exceptions (see below).

> For the `system` field GNU config and us already agree, right?

Without this PR?  No, not even close.

The only exceptions are listed above, and this PR puts the all in one place so it's easy to know "where do we behave differently than `gnu-config`"?  Fixing the first three would be a very jarring adjustment (and massive rebuild); I'm not opposed to that but if it is to be done it should happen in a separate PR after this one.  The last exception was a deliberate choice by nixpkgs (@sternenseemann).  The `vc4-elf` case really can't be fixed.

amjoseph-nixpkgs replied:
```
$ /nix/store/vqs7rm7vn8v3ja3sar609zlg6xd2zczh-gnu-config-2023-01-21/config.sub x86_64-linux
x86_64-pc-linux-gnu
```

```
$ nix-instantiate --eval --expr 'with (import ./lib).systems; (elaborate { system = "x86_64-linux"; }).config'
"x86_64-unknown-linux-gnu"
```

amjoseph-nixpkgs commented on lib/tests/triples.nix:
```diff
@@ -0,0 +1,157 @@
+      (builtins.tryEval (canonicalize triple))
+        .success))
+
+    #
+    # The following are triples which we have deliberately
+    # chosen to accept differently than gnu-config.  Entries here
+    # impose an ongoing burden on nixpkgs, so PRs which add new
+    # triples here should allow for a longer review period and
+    # should be advertised widely.
+    #
+    (lib.filter (v:
+
+      # gnu-config infers the vendor "pc" for these Nix doubles, but
+      # nixpkgs has used ${cpu}-unknown-linux-gnu for a very long time.
+      v != "x86_64-linux" && v != "i686-linux" &&
+      v != "x86_64-linux-gnu" && v != "i686-linux-gnu" &&
+      v != "x86_64-darwin" && v != "aarch64-darwin" &&
+
+      # We have been shipping Broadcom VC4 forks of gcc/binutils
+      # using nonstandard triples since 4aa1ffae041bb9c65eb3067e9dbaaa70710ed100
+      (canonicalize v) != "vc4-elf"
+
+      # This triple is special to GHC/Cabal/GHCJS and not recognized by autotools
+      # See: https://gitlab.haskell.org/ghc/ghc/-/commit/6636b670233522f01d002c9b97827d00289dbf5c
+      # https://github.com/ghcjs/ghcjs/issues/53
+      && (canonicalize v) != "javascript-unknown-ghcjs"
```
The only exceptions are listed above, and this PR puts the all in one place so it's easy to know "where do we behave differently than `gnu-config`"?

- Fixing the first group (six cases) would be a very jarring adjustment (and massive rebuild); I'm not opposed to that but if it is to be done it should happen in a separate PR after this one.
- The last exception was a deliberate choice by nixpkgs (@sternenseemann).
- The `vc4-elf` case really can't be fixed, `gnu-config` rejects the triple chosen by the upstream package.

amjoseph-nixpkgs replied:
```
$ /nix/store/vqs7rm7vn8v3ja3sar609zlg6xd2zczh-gnu-config-2023-01-21/config.sub x86_64-linux
x86_64-pc-linux-gnu
```

```
$ nix-instantiate --eval --expr \
  'with (import ./lib).systems; (elaborate { system = "x86_64-linux"; }).config'
"x86_64-unknown-linux-gnu"
```

amjoseph-nixpkgs replied:
If GNU config applies the following patch, we can drop the third exception:

- [config-patches: [PATCH] config.sub: add javascript-unknown-ghcjs
](https://lists.gnu.org/archive/html/config-patches/2023-08/msg00010.html)

Ericson2314 replied:
Upstream GHC is moving away from this one, so I don't feel so eager to upstream it. It was a temp hack I made.

amjoseph-nixpkgs reviewed (commented):
I probably should have titled this PR "add test suite **and fix our broke-ass parser so it passes the tests**".

This was a lot of work.  Very boring and tedious work.

> The general principle for me is that it is fine for either us or GNU config to reject something the other accepts, but they shouldn't both accept something and normalize it in a different way.

That is what this PR does, with three clearly enumerated exceptions (see below).

> For the `system` field GNU config and us already agree, right?

Without this PR?  No, not even close.

The most glaring disagreement is nixpkgs' most-popular platform: `x86_64-linux`!

```
$ /nix/store/vqs7rm7vn8v3ja3sar609zlg6xd2zczh-gnu-config-2023-01-21/config.sub x86_64-linux
x86_64-pc-linux-gnu

$ nix-instantiate --eval --expr \
  'with (import ./lib).systems; (elaborate { system = "x86_64-linux"; }).config'
"x86_64-unknown-linux-gnu"
```

Although nobody really cares what vendor you use in a triple, this points out one of the problems with the knee-jerk "we'll just canonicalize everything and the world will be happy".

These commits in this PR are needed in order to fix our parser in the sense of the general principle stated above:

```
lib/systems: add vendors.solo5
lib/systems: fix vendor fallback logic
lib/systems: fix abi inference for aarch32-linux
lib/systems: add vendors.{ibm,xilinx}
lib/systems: default arm-apple-linux to -gnu ABI
lib/systems: add and check valid kernels for each abi
lib/systems: handle Solaris properly
lib/systems: add kernels."", distinct from kernels.none
lib/systems: rename vendors.none -> vendors.""
lib/systems: replicate cygwin/msvc parsing hack when unparsing
```

The general principle you state is the same one I use, except:

- I don't think we should reject a triple that GNU config accepts unless it involves a coordinate (i.e. `cpu`, `vendor`, `kernel`, etc) that we reject unconditionally, like `amiga-*-*`, `*-motorola-*`, or `*-*-vxworks`.  Once we accept a coordinate, we shouldn't pick and choose which triples it's allowed in.
  - The only valid reason for rejecting a coordinate value should be "nobody's added it yet".  Now that we have a test suite, if somebody submits a PR to add a coordinate value and it passes the test, the "burden of proof" should be on anybody opposed to merging it.  The goal here is to outsource policy decisions as much as possible, so we make as few of them as possible.
- Rejecting a triple we *previously accepted* should use the same `lib.warn`-for-one-release-cycle approach taken by all other breaking changes to `lib`, unless the triple is so obviously bogus we can be sure nobody has ever used it, like `mmix-apple-windows-elf`"

amjoseph-nixpkgs commented:
> I think a good compromise then is simply exclude from the generated test cases anything where we disagree

Then what's the point of the tests?  They will pass trivially, because you're deliberately throwing out the cases where they fail.  Why bother running them?

Commit:
lib.systems.parse: rename abis.unknown to abis.""

Commit:
lib.systems: allow *-none-elf

Commit:
lib.systems: clean up abi defaulting

Artturin commented:
Rebased (to staging accidentally) and resolved conflict

Artturin commented:
`pkgsCross.mingwW64._7zz` which used to build, now fails at eval.

```
$ nix build ".#pkgsCross.mingwW64._7zz"
error:
       … while evaluating a branch condition

         at /home/artturin/nixgits/nixpkgs/pkgs/stdenv/booter.nix:64:9:

           63|       go = pred: n:
           64|         if n == len
             |         ^
           65|         then rnul pred

       … while calling the 'length' builtin

         at /home/artturin/nixgits/nixpkgs/pkgs/stdenv/booter.nix:62:13:

           61|     let
           62|       len = builtins.length list;
             |             ^
           63|       go = pred: n:

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: kernel windows does not allow abi gnu
```

Commit:
lib.systems.parse: abis.gnu: allow kernels.windows

Commit:
lib.systems.parse: remove assertion now that gnu-config has bumped

amjoseph-nixpkgs commented:
> `pkgsCross.mingwW64._7zz` which used to build, now fails at eval.

Yeah, rebasing past the `gnu-config` bump caused that.  The `gnu-config` commit which caused it was partially reverted upstream, so I've bumped `gnu-config` again in https://github.com/NixOS/nixpkgs/pull/262682 in order to pick that up.  Will rebase once https://github.com/NixOS/nixpkgs/pull/262682 is merged and then this should be good again.

Commit:
lib.systems.inspect: exclude kernel="" from isNone

Commit:
lib.systems.parse: do not infer vendor="pc" for *-elf

amjoseph-nixpkgs commented:
Rebased on https://github.com/NixOS/nixpkgs/pull/262682

Commit:
lib.tests.systems: include lib.systems.doubles.vc4

Commit:
lib.systems.triples: remove exception for ghcjs-javascript (gnu-config now understands it)

gnu-config commit 6c78704d542cebfb56d17474fe9f8395e9defb94 added
upstream support for ghc's javascript triple, so we can remove the
exception now.

amjoseph-nixpkgs commented:
Upstream commit 947d419372739faceb322ff126c14ef98f67ea9a allows us to drop one of the three exceptions, so now only two remain:

- `vc4-elf`
- nixpkgs canonicalizes `""-linux` to `unknown-linux` instead of `pc-linux`

lovesegfault commented on lib/systems/parse.nix:
```diff
@@ -520,7 +520,7 @@ rec {
         vendor = vendor.name;
         kernel = kernelName kernel;
-        inherit optExecFormat;
+        #inherit optExecFormat;
```
Why doesn't this commit remove the line?

amjoseph-nixpkgs replied:
Thanks! 62aaddf904bd

lovesegfault commented on lib/tests/triples.nix:
```diff
@@ -0,0 +1,157 @@
+#
+# 1. has an executable (i.e. unambiguous) specification
+# 2. covers a superset of the systems on which nixpkgs could be useful
+# 3. is *not* under nixpkgs' control (bickering/drama avoidance)
```
I don't think this point is worth retaining in the comment.

amjoseph-nixpkgs replied:
Okay, I've pushed 758723a6e32c which tries to use less-negative language.

I do think that it's worth mentioning this feature.  I consider it to be a positive aspect of adopting some existing taxonomy rather than cooking up our own.

Ericson2314 commented:
I would still like to change configs in `examples.nix` so that we need less special logic, and view this as a change within our rights to do so.

lovesegfault commented:
Really cool work :)

Ericson2314 commented:
We could submit an upstream patch for `vc4`, for example.

I think we should try to merge the commits that don't add more special-casing first.

Ericson2314 commented:
https://lists.gnu.org/archive/html/config-patches/2023-10/msg00009.html Here is a patch for VideoCore 4

Ericson2314 commented:
https://github.com/NixOS/nixpkgs/pull/247490#issuecomment-1783927337 I left a long message on that PR that equally applies to this one.

In short, I think in the current form this PR adds to much complexity. However, if do not expand the set of normal forms as much, we an still achieve the overall goal of testing and comparability with a subset of GNU config. It's going to be a subset of what GNU config supports either way, so I think that is fine.

Commit:
lib/systems/parse.nix: remove commented-out obsolete attribute

Commit:
lib/tests/triples.nix: reword comment

Commit:
lib/tests/triples.nix: add --silent to make invocation

amjoseph-nixpkgs commented:
d474c87aff678090f108a23f0b3e521ae0d4e034 has landed on `master` so I can now switch this back from `staging`...

Commit:
lib/tests/triples.nix: add *-*-uefi exception

Commit:
Revert "lib/tests/triples.nix: add *-*-uefi exception"

This reverts commit 20a384ee5c642021786bb5981bb613d95273bb6e.

amjoseph-nixpkgs commented:
Fixed merge conflict.

Commit:
lib/systems/default.nix: fix darwin missing bootstrap url for platform x86_64-apple-darwin-

amjoseph-nixpkgs commented:
Hrm, I don't think this is my fault:

```
ran test tests/add.sh... [FAIL]
    ++ nix-store --add ./dummy
    + path1=/build/nix-test/add/store/n7knm3rma6fckrfwkx2vkiql3d1rgs1i-dummy
    + echo /build/nix-test/add/store/n7knm3rma6fckrfwkx2vkiql3d1rgs1i-dummy
    /build/nix-test/add/store/n7knm3rma6fckrfwkx2vkiql3d1rgs1i-dummy
    ++ nix-store --add-fixed sha256 --recursive ./dummy
    + path2=/build/nix-test/add/store/g1qxg63zbanhv79ibby90311521d4237-dummy
    + echo /build/nix-test/add/store/g1qxg63zbanhv79ibby90311521d4237-dummy
    /build/nix-test/add/store/g1qxg63zbanhv79ibby90311521d4237-dummy
    + test /build/nix-test/add/store/n7knm3rma6fckrfwkx2vkiql3d1rgs1i-dummy '!=' /build/nix-test/add/store/g1qxg63zbanhv79ibby90311521d4237-dummy
    + echo 'nix-store --add and --add-fixed mismatch'
    nix-store --add and --add-fixed mismatch
    + exit 1
make: *** [mk/lib.mk:128: tests/add.sh.test] Error 1
make: *** Waiting for unfinished jobs....
```

x10an14 commented:
Hey! Sorry to see this closed, was looking forward to seeing this in action =/

Can I inquire as to what's the plan going forward on this? Is this abandoned/halted for some (afaict) unspecified reason?

RaitoBezarius commented:
This is related to https://github.com/NixOS/nixpkgs/issues/50105#issuecomment-1905292471 and I suppose someone will have to pick the work to get it to finish line themselves.
